tests/k_winners_test.py

```diff
@@ -38,210 +38,109 @@ class KWinnersTest(unittest.TestCase):
-
-        expected = torch.zeros_like(x)
-        expected[0, 0, 1, 0] = 1.1
-        expected[0, 0, 1, 1] = 1.2
-        expected[0, 1, 0, 1] = 1.2
-        expected[0, 2, 1, 0] = 1.3
-        expected[1, 0, 0, 0] = 1.4
-        expected[1, 1, 0, 0] = 1.5
-        expected[1, 1, 0, 1] = 1.6
-        expected[1, 2, 1, 1] = 1.7
-
-        self.assertEqual(result.shape, expected.shape)
-
-        num_correct = (result == expected).sum()
-        self.assertEqual(num_correct, result.reshape(-1).size()[0])
-
-        indices = ctx.saved_tensors[0]
-        expected_indices = torch.tensor([[2, 3, 10, 5], [0, 4, 5, 11]])
-        num_correct = (indices == expected_indices).sum()
-        self.assertEqual(num_correct, 8)
+        self.assertEqual(len(grad_x.nonzero()), len(expected.nonzero()))
 
-        # Test that gradient values are in the right places, that their sum is
-        # equal, and that they have exactly the right number of nonzeros
-        out_grad, _, _, _ = F.KWinners2d.backward(ctx, self.gradient2)
-        out_grad = out_grad.reshape(2, -1)
-        in_grad = self.gradient2.reshape(2, -1)
-        self.assertEqual((out_grad == in_grad).sum(), 8)
-        self.assertEqual(len(out_grad.nonzero()), 8)
-
-    def test_four(self):
-        """Equal duty cycle, boost factor=0, k=3, batch size=2."""
+    def test_k_winners_module(self):
         x = self.x2
 
-        ctx = TestContext()
-
-        result = F.KWinners2d.forward(ctx, x, self.duty_cycle, k=3, boost_strength=0.0)
+        n = 6
 
-        expected = torch.zeros_like(x)
-        expected[0, 0, 1, 1] = 1.2
-        expected[0, 1, 0, 1] = 1.2
-        expected[0, 2, 1, 0] = 1.3
-        expected[1, 1, 0, 0] = 1.5
-        expected[1, 1, 0, 1] = 1.6
-        expected[1, 2, 1, 1] = 1.7
-
-        self.assertEqual(result.shape, expected.shape)
-
-        num_correct = (result == expected).sum()
-        self.assertEqual(num_correct, result.reshape(-1).size()[0])
-
-        indices = ctx.saved_tensors[0]
-        expected_indices = torch.tensor([[3, 10, 5], [4, 5, 11]])
-        num_correct = (indices == expected_indices).sum()
-        self.assertEqual(num_correct, 6)
-
-        # Test that gradient values are in the right places, that their sum is
-        # equal, and that they have exactly the right number of nonzeros
-        out_grad, _, _, _ = F.KWinners2d.backward(ctx, self.gradient2)
-        out_grad = out_grad.reshape(2, -1)
-        in_grad = self.gradient2.reshape(2, -1)
-        self.assertEqual((out_grad == in_grad).sum(), 6)
-        self.assertEqual(len(out_grad.nonzero()), 6)
-
-    def test_k_winners2d_module(self):
-        x = self.x2
-
-        kw = KWinners2d(
+        kw = KWinners(
+            n,
             percent_on=0.333,
-            channels=3,
             k_inference_factor=0.5,
             boost_strength=1.0,
             boost_strength_factor=0.5,
             duty_cycle_period=1000,
         )
 
+        # Expect 2 winners per batch (33% of 6)
```

Comment:
I was initially a bit confused. Shouldn't the number of winners be 0.5 * 0.333 * 6 since k_inference_factor=0.5

Is kw.training True here by default? Maybe we should be more explicit.

Reply:
True, that is not clear. Will set the module explicitly to train()